docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,140 @@
+   in new clusters
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to use the assisted service to repurpose hosts 
+from existing clusters without requiring access to the host management 
+infrastructure (BMC, hypervisor, etc).
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. A way to query for URLs to the iPXE boot artifacts over the REST API should be
+implemented. Currently these URLs are only exposed through the iPXE script and
+in the InfraEnv CR, but the agent running on a node will need access to these
+artifacts.
+  * This can be added to the existing payload returned when querying an InfraEnv
+  * Alternatively this would need a new API endpoint
+    * `/v2/infra-envs/{infra_env_id}/boot-artifacts`?
+2. The agent will likely require two changes/additions
+  * A step to query the boot artifacts and write the boot entry
+  * A step to reboot the host into the new boot entry
+3. The service will need a new API to prepare a cluster for hosts to be reclaimed
+  * This needs to be separate from unbind so that the user has time to cleanly
+    remove a host from the cluster
+  * `/v2/clusters/{cluster_id}/prepare-unbind`?
+
+#### Assumptions
+
+This approach only works if the host is currently part of a working cluster
+which the assisted service can access using the kubeconfig generated during
+install.
+
+There will also need to be enough space on the `/boot` partition to download the
+initrd and kernel, but the agent can check for this during the preparation step.
+
+### Risks and Mitigations
+
+It's possible that something could go wrong during the boot process. A user 
+that doesn't have access to the host infrastructure wouldn't be able to debug 
+or solve this type of problem. In this case the user is no worse off than they
+currently would be and will have to boot the discovery ISO manually.
+
+## Design Details [optional]
+
+### Open Questions
+
+1. Will users want to reclaim individual hosts in an existing cluster or will
+   the whole cluster always be decommissioned at the same time?
+  * This will determine how we target the MachineConfig
+  * Ideally we don't want to be running an agent on nodes we know we don't want
+    to reclaim
+2. How long is the kubeconfig for a particular cluster saved?
+  * It needs to still be around whenever a user wants to reclaim hosts
+
+### UI Impact
+
+A UI would need to be developed for this feature, but it won't be used in the 
+cloud so it doesn't need a UI there.
+
+### Test Plan
+
+## Drawbacks
+
+This will only work in environments where the service has access to the
+kubernetes API on the cluster. This specifically means that it won't be useful
+in the SaaS.
+
+## Alternatives
+
+The only strategy which would work in the cloud would be a full polling 
+approach where we set up an agent on the node at install that constantly asks
+the service if it should reprovision.
+
+This is likely not worth the resources and clusters don't exist long enough in
```

Comment:
It also sounds like a security nightmare. I wouldn't want my AI installed clusters to be potentially controlled by the installation service that helped me install it, it feels wrong

Reply:
I mean that's what we're doing anyway though, right?
We're still using the kubeconfig that we generated on install to execute privileged operations on your machines from the installer.

We're just only running the agent when you tell us to rather than running it the entire time.

Reply:
As mentioned in other threads this is never going to be relevant in the SaaS as there's no reason to assume we can access the kubernetes API server.